# Reviewer Comments
 
- [ ] 1. Separate INDEL analysis into insertions and deletions
- [ ] 2. Consider DFE for different lengths
    - [ ] run model per length bin, maybe 1+2bp, 3bp, 4+5bp, 6bp? using the a class model
- [ ] 3. run with high classes c = 10
    - [ ] see if power analysis in Kousathanas and Keightley 2013 or Keightley and Eyre-Walker 2010
    - [ ] if not do power analysis, simulate 2 class data with increasing pop size, see how often 1 class model is rejected at each sample size
- [ ] 4. table 4 statistics on nonsense mutations would be a good comparison
    - [ ] pi for premature stop codons, 0fold only to life easier, ignore MNPs
- [x] 5. Small contrast between frame shift and in frame, yet frame shift more disruptive
    - [x] frame shift occur at higher frequency and are more deleterious, in frame occur at lower frequency but persist longer, result is that estimates are closer together
- [ ] 6. Allele frequency distribution for INDELs
    - [ ] add AR
- [ ] 7. Mutation rate influence pol error rate, yet neutral ref doesn't all pol error to vary
- [ ] 8. Discuss that insertions look neutral but dels don't and compare to Leushkin et al.
- [ ] 9. Synonymous as neutral ref
    - [ ] could use fourfold or ARs
    - [ ] check length distribution of AR INDELs
- [ ] 10. Page 2 column 2 inherit should be inherent
- [ ] 11. Page 2 column 2 start of paragraph 'An additional challenge...' is poorly written
 